File: src/leetcode/python/hard/InsertDeleteGetRandomO1DuplicatesAllowed_381.py
import collections
import random


class InsertDeleteGetRandomO1DuplicatesAllowed_381(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.index = []
        self.value = collections.defaultdict(set)

    def insert(self, val):
        """
        Inserts a value to the collection. Returns true if the collection did not already contain the specified element.
        :type val: int
        :rtype: bool
        """
        self.index.append(val)
        self.value[val].add(len(self.index) - 1)
        return len(self.value[val]) == 1

    def remove(self, val):
        """
        Removes a value from the collection. Returns true if the collection contained the specified element.
        :type val: int
        :rtype: bool
        """
        if self.value[val]:
            idx = self.value[val].pop()
            val = self.index[-1]
            self.index[idx] = val
            if self.value[val]:
                self.value[val].add(idx)
                self.value[val].discard(len(self.index)-1)
            self.index.pop()
            return True
        return False

    def getRandom(self):
        """
        Get a random element from the collection.
        :rtype: int
        """
        return random.choice(self.index)

File: src/leetcode/java/InsertDeleteGetRandomO1DuplicatesAllowed_381.java
package leetcode.java;

import java.util.HashMap;
import java.util.Random;
import java.util.TreeSet;

public class InsertDeleteGetRandomO1DuplicatesAllowed_381 {
  private HashMap<Integer, TreeSet<Integer>> value;
  private HashMap<Integer, Integer> index;
  private Random random;

  /** Initialize your data structure here. */
  public InsertDeleteGetRandomO1DuplicatesAllowed_381() {
    this.value = new HashMap<>();
    this.index = new HashMap<>();
    this.random = new Random();
  }

  /**
   * Inserts a value to the collection. Returns true if the collection did not already contain the
   * specified element.
   */
  public boolean insert(int val) {
    boolean contain = value.containsKey(val);
    if (!contain) value.put(val, new TreeSet<>());
    TreeSet<Integer> idxSet = value.get(val);
    idxSet.add(index.size());
    index.put(index.size(), val);
    return !contain;
  }

  /**
   * Removes a value from the collection. Returns true if the collection contained the specified
   * element.
   */
  public boolean remove(int val) {
    if (!value.containsKey(val)) return false;
    TreeSet<Integer> idxSet = value.get(val);
    int idx = idxSet.pollLast();
    if (idxSet.size() == 0) value.remove(val);
    int exchangeKey = index.get(index.size() - 1);
    if (exchangeKey == val) {
      index.remove(index.size() - 1);
      return true;
    }
    int exchangeidx = value.get(exchangeKey).pollLast();
    index.put(idx, exchangeKey);
    value.get(exchangeKey).add(idx);
    index.remove(exchangeidx);
    return true;
  }

  /** Get a random element from the collection. */
  public int getRandom() {
    return index.get(random.nextInt(index.size()));
  }
}
